### PR TITLE
ensure consistency for the generated image

### PR DESCRIPTION
**Pull request by christian-heusel**
This just applies the proposed fix from @Artefact2 in the linked issue.
Should we make this behaviour more configurable as it increases the space requirements, which could be a breaking change for people with full `/boot`?

fixes https://github.com/archlinux/mkinitcpio/issues/64

### CODE REVIEW

**Comment by christian-heusel** [comment]
I could also add Co-Authorship for the person that opened the issue if wanted!

**Comment by Artefact2** [comment]
It's a trivial fix, I don't mind. Happy to see it fixed in mainline :)

**Review by Foxboron — APPROVED** [review]

**Comment by nl6720** [comment]
> it increases the space requirements

IMHO this is quite a big issue. People with 100 MiB `/boot` already exceed space requirements frequently enough.

The [suggested `/boot` size in the wiki](https://wiki.archlinux.org/title/Partitioning#/boot) would need updating too.

**Comment by christian-heusel** [comment]
> it increases the space requirements

It increases the space requirements by `max(img_filesizes)` as the rename is happening after each image, so we get to a new needed minimum amount of `sum(img_filesizes) + max(img_filesizes)`.
For the two machines i am testing this with this means the increase is +34M because of the size of `initramfs-linux-fallback.img`.

How big can these get?

**Comment by nl6720** [comment]
Generally, the more crap you add, the bigger it gets.
I myself have a 49 MiB `initramfs-linux-fallback.img`, but I'm using 
```bash
COMPRESSION="xz"
COMPRESSION_OPTIONS=(-9e)
```
With the default compressions options, it's 66 MiB for me.

**Comment by Artefact2** [comment]
The initramfses can get quite large, especially in NVIDIA users that want early kms (80-ish megs last time I checked).

**Review by nl6720 — CHANGES_REQUESTED** [review]
I think there should be a fallback to truncating in case there's not enough free space.

File: mkinitcpio [hunk 1]
```diff
@@ -228,6 +228,18 @@ build_image() {
...
+    if [[ -f "$out" ]]; then
+        local curr_size space_left_on_device
+
+        curr_size=$(stat --format="%s" $out)
+        space_left_on_device=$(($(stat -f --format="%a*%S" $out)))
+
+        # check if there is enough space on the device to write the image to a tempfile, fallback otherwise
+        if (( $(($curr_size*2)) < $space_left_on_device )); then
+            compressout="$out".tmp
+        fi
```
Comment (nl6720) [inline mkinitcpio:240]:
```suggestion
        (( $((curr_size*2)) < space_left_on_device )) && compressout="$out".tmp
```

File: mkinitcpio [hunk 2]
```diff
@@ -228,6 +228,18 @@ build_image() {
             ;;
     esac
 
+    if [[ -f "$out" ]]; then
+        local curr_size space_left_on_device
+
+        curr_size=$(stat --format="%s" $out)
+        space_left_on_device=$(($(stat -f --format="%a*%S" $out)))
```
Comment (nl6720) [inline mkinitcpio:235]:
```suggestion
        curr_size="$(stat --format="%s" "$out")"
        space_left_on_device="$(($(stat -f --format="%a*%S" "$out")))"
```

File: mkinitcpio [hunk 3]
```diff
@@ -238,7 +250,13 @@ build_image() {
             sort -z |
             LANG=C bsdtar --uid 0 --gid 0 --null -cnf - -T - |
             LANG=C bsdtar --null -cf - --format=newc @- |
-            $compress "${COMPRESSION_OPTIONS[@]}" > "$out"
+            $compress "${COMPRESSION_OPTIONS[@]}" > "$compressout"
+
+    # sync and rename as we wrote to a tempfile
+    if [[ "$compressout" != "$out" ]]; then
+        sync -d "$compressout"
+        mv -f "$compressout" "$out"
```
Comment (nl6720) [inline mkinitcpio:258]:
```suggestion
        sync -d -- "$compressout"
        mv -f -- "$compressout" "$out"
```

**Review by nl6720 — CHANGES_REQUESTED** [review]

File: mkinitcpio [hunk 4]
```diff
@@ -228,6 +228,18 @@ build_image() {
...
     esac
 
+    if [[ -f "$out" ]]; then
+        local curr_size space_left_on_device
+
+        curr_size=$(stat --format="%s" $out)
+        space_left_on_device=$(($(stat -f --format="%a*%S" $out)))
+
+        # check if there is enough space on the device to write the image to a tempfile, fallback otherwise
+        if (( $(($curr_size*2)) < $space_left_on_device )); then
```
Comment (nl6720) [inline mkinitcpio:238]: This probably needs to a little buffer.

**Review by nl6720 — CHANGES_REQUESTED** [review]

**Comment by nl6720** [comment]
I'm getting:
```
==> ERROR: Image generation FAILED: find reported an error
```
The `find` in question is this one:

```bash
    find . -mindepth 1 -printf '%P\0' |
            sort -z |
            LANG=C bsdtar --uid 0 --gid 0 --null -cnf - -T - |
            LANG=C bsdtar --null -cf - --format=newc @- |
            $compress "${COMPRESSION_OPTIONS[@]}" > "$compressout"
```

Which turns into:

```bash
+ find . -mindepth 1 -printf '%P\0'
+ sort -z
+ LANG=C
+ bsdtar --uid 0 --gid 0 --null -cnf - -T -
+ xz --check=crc32 -9e
+ LANG=C
+ bsdtar --null -cf - --format=newc @-
+ [[ /boot/initramfs-linux.img != \/\b\o\o\t\/\i\n\i\t\r\a\m\f\s\-\l\i\n\u\x\.\i\m\g ]]

```

This is with a 100MiB `/boot`:

```
Filesystem     Type  Size  Used Avail Use% Mounted on
/dev/loop0     vfat   99M   81M   18M  83% /boot
```

**Comment by nl6720** [comment]
> ```
> ==> ERROR: Image generation FAILED: find reported an error
> ```

This issue is now gone.

**Comment by nl6720** [comment]
Now it's just a question of how much space to reserve. :thinking:

**Comment by christian-heusel** [comment]
> Now it's just a question of how much space to reserve. thinking

Wdym? Imo we could just leave it as is and maybe emit a warning if the partition is to full 🤔 But since its just not doing a more safe upgrade I think this also could be left as-is.

> The [suggested /boot size in the wiki](https://wiki.archlinux.org/title/Partitioning#/boot) would need updating too.

With the now implemented behaviour (doing atomic rename when possible, otherwise use the current behaviour) I think one could also argue that there is no change needed in the wiki.

File: mkinitcpio [hunk 5]
```diff
@@ -228,6 +228,17 @@ build_image() {
...
     esac
 
+    if [[ -f "$out" ]]; then
+        local curr_size space_left_on_device
+
+        curr_size="$(stat --format="%s" "$out")"
+        space_left_on_device="$(($(stat -f --format="%a*%S" "$out")))"
+
+        # check if there is enough space on the device to write the image to a tempfile, fallback otherwise
+        # this assumes that the new image is not more than 1 1/4x the size of the old one
```
Comment (nl6720) [inline mkinitcpio:238]:
```suggestion
        # this assumes that the new image is not more than 1¼ times the size of the old one
```

**Comment by nl6720** [comment]
> With the now implemented behaviour (doing atomic rename when possible, otherwise use the current behaviour) I think one could also argue that there is no change needed in the wiki.

:+1:

**Comment by nl6720** [comment]
> Wdym? Imo we could just leave it as is and maybe emit a warning if the partition is to full thinking But since its just not doing a more safe upgrade I think this also could be left as-is.

I mean, is 1¼× enough to account for the new image's potential size increase?

**Comment by christian-heusel** [comment]
> I mean, is 1¼× enough to account for the new image's potential size increase?

🤷🏻‍♂️ I dont know, for my current setup this would mean 8.5MB slack for the fallback img. I think it should be enough for regular upgrades of any kind.

> maybe emit a warning if the partition is to full thinking

what do you think about this?

**Review by nl6720 — APPROVED** [review]

**Comment by nl6720** [comment]
It'd feel safer to additionally add some fixed (not dependent on the previous image size) amount of free space to the equation.

**Comment by christian-heusel** [comment]
Re-requesting review because of the bigger changes after the initial approve 😊

**Comment by Foxboron** [comment]
I need to test it so I understand how it works. It looks good to me but please squash the last 3 commits into the 2nd one, or squash everything together :)

**Comment by christian-heusel** [comment]
@Foxboron commits are now squashed 👍🏻

**Pull request merged**